[Test] Goal 테스트코드 H2 DB 활용 방식 리팩토링(ALL)

rlqor9688 opened the pull request:
## 📌 관련 이슈
- Closes #74

## ✨ 변경 사항
- H2 DB를 활용한 테스트코드 리팩토링 적용 1차
- 로직이 단순한 Fetcher와 일부 서비스 코드에 H2 DB를 활용한 테스트코드를 적용했습니다.

[Fetcher]
1. GoalDetailFetcherFactoryTest
2. GoalEtcFetcherTest
3. GoalHousingFetcherTest
4. GoalRetirementFetcherTest
>> H2 DB로 테스트코드 리팩토링 완료(기존 테스트코드 남겨 놓음)

[Service]
1. RetirementReferenceValueServiceTest 
2. GoalRetirementCalculateServiceTest 
>> H2 DB로 테스트코드 리팩토링 완료(기존 테스트코드 남겨 놓음)
3. OpenApiApartmentPriceServiceTest : 예외 상황 검증 테스트코드 추가 (이 서비스는 Repository를 사용하지 않아 H2 DB 리팩토링X)
4. DefaultRetirementGoalServiceTest
5. GoalServiceImplTest

## 📷 Postman 
- H2GoalDetailFetcherFactoryTest
![image](https://github.com/user-attachments/assets/d9ef0225-0ebe-4309-a8a4-e1fe5cb0825f)

- H2GoalEtcFetcherTest
![image](https://github.com/user-attachments/assets/b3578404-cac8-43d6-b97f-a8197fa929d6)

- H2GoalHousingFetcherTest
![image](https://github.com/user-attachments/assets/5ab5a2b6-917d-4b4d-9725-7dfbe9008c00)

- H2GoalRetirementFetcherTest
![image](https://github.com/user-attachments/assets/4845de81-20b9-45c6-96ba-b2103324cd20)

- H2GoalRetirementCalculateServiceTest
![image](https://github.com/user-attachments/assets/5920c47e-da34-45bc-ba39-2769ab573421)

- H2RetirementReferenceValueServiceTest
![image](https://github.com/user-attachments/assets/3d5cb80a-fa25-45f4-95ef-0e4d21a88dce)

- H2DefaultRetirementGoalServiceTest
![image](https://github.com/user-attachments/assets/8d81ad6a-500b-4d1c-924d-054e5e9345ca)

- H2GoalServiceImplTest
![image](https://github.com/user-attachments/assets/af32d26f-4a5a-433b-91c9-42f05ac6775f)

## ✅ 체크리스트
- [ ] 관련 이슈에 연결됨
- [ ] 기능이 정상적으로 동작함
- [ ] 코드 리뷰를 받았음
- [ ] 불필요한 로그/코드 제거됨

## 💬 기타 참고 사항
- 테스트 방법, 추가 설명 등 필요한 경우 작성해 주세요.

rlqor9688 committed:
[refactor] H2 테스트 코드 리팩토링
[Fetcher]
1. GoalDetailFetcherFactoryTest
2. GoalEtcFetcherTest
3. GoalHousingFetcherTest
4. GoalRetirementFetcherTest
5. RetirementReferenceValueServiceTest
H2 DB로 테스트코드 리팩토링 완료(기존 테스트코드 남겨 놓음)

rlqor9688 committed:
[fix] 예외 상황 검증 로직 추가
OpenApiApartmentPriceServiceTest : 예외 상황 검증 테스트코드 추가

rlqor9688 committed:
[refactor] GoalRetirementCalculateServiceTest
H2 DB로 테스트 코드 리팩토링

rlqor9688 committed:
[docs] 줄맞춤, 중복 수정

rlqor9688 committed:
[fix] User 객체 생성 시 NotNull인 필드 추가

rlqor9688 committed:
[refactor] H2 DB 테스트코드 구현 Update
1. H2GoalDetailFetcherFactoryTest
2. H2DefaultRetirementGoalServiceTEst
3. H2GoalServiceImplTest

rlqor9688 committed:
[feat] GoalEntity 내 메서드 추가

rlqor9688 committed:
[fix] 테스트 클래스 전체 실행 시 Entity 중복 이슈 해결
테스트 클래스 전체 실행 시 H2 DB에 데이터가 저장되면서 유저 닉네임, 이메일 중복 발생하는 문제 해결

byeongtaek12 reviewed (approved):
확인 완료했고 수고하셨습니다. 기존 pr한 테스트 가지고 리팩토링 진행하신 거라 코드를 하나하나 뜯어보진 않았고 H2 DB는 이런 구조로 할 수 있구나라는 것만 봤습니다. DB에 잘 뜨던가요? save() 한 것들이?

ChangHoonS commented on src/test/java/org/example/lifechart/domain/goal/fetcher/H2GoalDetailFetcherFactoryTest.java:
@SpringBootTest는 Spring Context 전체를 로딩해서 @Component, @Service, @Repository 같은 Bean을 자동 주입한다고 합니다.
그런데 @BeforeEach는 빈을 수동 생성 하는 것이라 Spring이 주입한 것을 덮어써버린다고 합니다.
그래서 의도한대로 동작하지 않거나 Context 충돌이 발생할 수도 있다고 하는데 우선 충돌 등 문제는 없어보이나
생각 해볼 필요가 있어보입니다.
찾아보기에는 
@Autowired
	private GoalDetailFetcherFactory goalDetailFetcherFactory; 부분이 있어서 

@BeforeEach
	void setup() {
		goalDetailFetcherFactory = new GoalDetailFetcherFactory(List.of(goalEtcFetcher, goalHousingFetcher, goalRetirementFetcher));
	}
	이 부분은 없어도 무방하다고 합니다. 참고 부탁드립니다.

rlqor9688 replied:
오 덕분에 깊이 있게 이해했습니다! 감사합니다!

ChangHoonS commented on src/test/java/org/example/lifechart/domain/goal/fetcher/H2GoalDetailFetcherFactoryTest.java:
그리고 DB를 사용해 테스트를 하는 상황이라 데이터간에 서로 영향을 줄 수 있어 내용을 바꾸라고 하셨는데
각 테스트 메서드 앞에 @Transactional을 사용하면 테스트 후에 롤백이 되서 격리가 가능하다고 합니다.
수정이 필요하다 보다는 알고계시면 좋을 것 같아서 적어둡니다.

rlqor9688 replied:
오~ 그렇군요! Test에서 Transactional을 붙일 수가 있군요! 덕분에 하나 제대로 배웠습니다! 감사합니다!

rlqor9688 replied:
말씀주신 내용 적용해보니 다 잘 되네요! 감사합니다! 적용해서 현재 PR에 push하겠습니다.

ChangHoonS replied:
넵 답변감사합니다!

rlqor9688 commented:
@byeongtaek12 네 DB를 직접 확인해보진 않았는데, 같은 닉네임으로 유저를 생성하려고 하면 중복 에러가 나오고, 클래스 전체를 실행했을 때 저장된 Goal들의 Goal 번호가 쌓여서 출력되더라고요! 우선 이렇게 간접적으로 확인했습니다. 직접 확인하는 방법도 한번 알아볼게요! 감사합니다!

rlqor9688 committed:
[fix] beforeEach 설정 삭제 & Transactional 시험 적용
- BeforeEach와 Autowired로 주입한 속성이 충돌을 일으킬 가능성이 있어, BeforeEach 설정 삭제
- @Transactional 설정 시 클래스 내 다수의 메서드를 동시에 실행시켜도 DB 중복 이슈가 발생하지 않는 상황 확인
